src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs

```diff
@@ -73,24 +81,40 @@ impl ReclusterMutator {
         std::mem::take(&mut self.selected_blocks)
     }
 
-    pub fn mutation_logs(&self) -> MutationLogs {
-        self.mutation_logs.clone()
-    }
-
     #[async_backtrace::framed]
     pub async fn target_select(
         &mut self,
-        blocks_map: BTreeMap<i32, Vec<(BlockMetaIndex, Arc<BlockMeta>)>>,
+        compact_segments: Vec<(SegmentLocation, Arc<CompactSegmentInfo>)>,
     ) -> Result<bool> {
+        let mut selected_segments = Vec::with_capacity(compact_segments.len());
+        let mut selected_indices = Vec::with_capacity(compact_segments.len());
+        let mut selected_statistics = Vec::with_capacity(compact_segments.len());
+        compact_segments.into_iter().for_each(|(loc, info)| {
+            selected_statistics.push(info.summary.clone());
+            selected_segments.push(info);
+            selected_indices.push(loc.segment_idx);
+        });
+
+        let blocks_map = self.gather_block_map(selected_segments).await?;
+
+        let mem_info = sys_info::mem_info().map_err(ErrorCode::from_std_error)?;
+        let max_memory_usage = self.ctx.get_settings().get_max_memory_usage()? as usize;
+        let memory_threshold =
+            cmp::min(mem_info.avail as usize * 1024, max_memory_usage) * 50 / 100;
+
+        let mut remained_blocks = Vec::new();
+        let mut selected = false;
         for (level, block_metas) in blocks_map.into_iter() {
-            if block_metas.len() < 2 {
+            let len = block_metas.len();
+            if selected || len < 2 {
```

Comment:
I think this logic is an error, remain_blocks should be the blocks which is not selected for this recluster ? Maybe I'm wrong.

Reply:
I'm wrong, sorry